rfcs/text/0012_encryption_key_rotation.md

```diff
@@ -0,0 +1,118 @@
+
+# Detailed design
+
+The core idea is that when the encryption key needs to be rotated then a new key is generated and becomes a primary one, and the old one moves to the `keyRotation` section:
+
+```yaml
+xpack.encryptedSavedObjects:
+  encryptionKey: "NEW-encryption-key"
+  keyRotation:
+    decryptionOnlyKeys: ["OLD-encryption-key"]
+```
+
+As the name implies, the key from the `decryptionOnlyKeys` is only used to decrypt content that we cannot decrypt with the primary encryption key. It's allowed to have multiple decryption-only keys at the same time. When user creates a new Saved Object or updates the existing one then its content is always encrypted with the primary encryption key. Config schema won't allow having the same key in `encryptionKey` and `decryptionOnlyKeys`.
+
+Having multiple decryption keys at the same time brings one problem though: we need to figure out which key to use to decrypt specific Saved Object. If our encryption keys could have a unique ID that we would store together with the encrypted data (we cannot use encryption key hash for that for obvious reasons) we could know for sure which key to use, but we don't have such functionality right now and it may not be the easiest one to manage through `yml` configuration anyway. 
+
+Instead, this RFC proposes to try available existing decryption keys one by one to decrypt Saved Object and always start from the primary one. This way we won't incur any penalty while decrypting Saved Objects that are already encrypted with the primary encryption key, but there will still be some cost when we have to perform multiple decryption attempts. See the [`Drawbacks`](#drawbacks) section for the details.
+
+Technically just having `decryptionOnlyKeys` would be enough to cover the majority of the use cases, but the old decryption-only keys should be eventually disposed. At this point administrators would like to make sure _all_ Saved Objects are encrypted with the new primary encryption key. Another reason to re-encrypt all existing Saved Objects with the new key at once is to preventively reduce the performance impact of the multiple decryption attempts.
+
+We'd like to make this process as simple as possible while meeting the following requirements:
+
+* It should not be required to restart Kibana to perform this type of migration since Saved Objects encrypted with the another encryption key can theoretically appear at any point in time.
+* It should be possible to integrate this operation into other operational flows our users may have and any user-friendly key management UIs we may introduce in this future.
+* Any possible failures that may happen during this operation shouldn't make Kibana nonfunctional.
+* Ordinary users should not be able to trigger this migration since it may consume a considerable amount of computing resources.
+
+We think that the best option we have right now is a dedicated API endpoint that would trigger this migration:
```

Comment:
There may also be the option of using a soon-to-be-developed CLI tool: https://github.com/elastic/kibana/issues/65788.

Reply:
Good point, wondering if it still can use the APIs we may expose.

Reply:
Good question! My guess would be that the CLI tool can't use the Kibana APIs as it would probably not be running. The tool could generate a new encryption key and move the old one to the `keyRotation` yml setting for the admin if ever that makes things easier.

The original requirement for the tool is to set an initial `xpack.encryptedSavedObjects.encryptionKey` value to facilitate setting up Kibana.

Reply:
Got it, thanks for the context! I'll try to sync with @tylersmalley next week then to make sure we're aligned.